File: MetaBrain/observability/performance-metrics.ts
import type { ObservabilityEvent } from "./types";

export function calculatePerformanceMetrics(events: ObservabilityEvent[]) {
  const latencies = events.map((event) => Number(event.payload_summary.latency_ms)).filter(Number.isFinite);
  return {
    measured_events: latencies.length,
    latency_avg_ms: latencies.length ? Math.round(latencies.reduce((sum, value) => sum + value, 0) / latencies.length) : 0,
    timeout_count: events.filter((event) => event.event_type.includes("timeout")).length,
  };
}

File: MB-Chat/observability_py/performance_metrics.py
from .types import ObservabilityEvent


def calculate_performance_metrics(events: list[ObservabilityEvent]) -> dict[str, int]:
    latencies = [float(event.payload_summary["latency_ms"]) for event in events if isinstance(event.payload_summary.get("latency_ms"), (int, float))]
    return {
        "measured_events": len(latencies),
        "latency_avg_ms": round(sum(latencies) / len(latencies)) if latencies else 0,
        "timeout_count": len([event for event in events if "timeout" in event.event_type]),
    }
